stablehlo/reference/Element.cpp

```diff
@@ -131,6 +131,35 @@ Element mapWithUpcastToDouble(const Element &el, FloatFn floatFn,
                                      debugString(type).c_str()));
 }
 
+template <typename T>
+std::enable_if_t<std::is_floating_point<T>::value, bool> areApproximatelyEqual(
+    T x, T y) {
+  // The machine epsilon has to be scaled to the magnitude of the values used,
+  return std::fabs(x - y) <=
+             std::numeric_limits<T>::epsilon() * std::fmax(x, y) ||
+         // unless the result is subnormal
+         std::fabs(x - y) < std::numeric_limits<T>::min();
+}
+
+// Checks if two APFloat values, f and g, are almost equal.
+bool areApproximatelyEqual(APFloat f, APFloat g) {
+  if (&f.getSemantics() != &g.getSemantics()) return false;
+
+  llvm::APFloatBase::cmpResult cmpResult = f.compare(g);
+  if (cmpResult == APFloat::cmpEqual) return true;
+  if (cmpResult == APFloat::cmpUnordered) return f.isNaN() == g.isNaN();
+  if (!f.isFiniteNonZero() || !g.isFiniteNonZero()) return false;
```

Comment:
Let me explain the rationale behind the way the function is implemented. The implementation is inspired from [APFLoat::compare](https://github.com/llvm/llvm-project/blob/1c2280264058c26aee2bcd8fe4ca90d0f843cc5c/llvm/lib/Support/APFloat.cpp#L2327)

I wanted to handle normal values separately for "almost equality" by calling  `areApproximatelyEqual<double> or areApproximatelyEqual<float>`, as shown in the code starting with comment "Both f and g are normal values.". Before doing that  I would like to handle all the non-normal cases. This is the same approach as how `APFloat::compare` is implemented. For example, before the [program point](https://github.com/llvm/llvm-project/blob/1c2280264058c26aee2bcd8fe4ca90d0f843cc5c/llvm/lib/Support/APFloat.cpp#L2376) , it handles all the  non-normal cases before handling the normal values. 

From the   `APFloat::compare`  implementation,  `cmpResult` equals to `APFloat::cmpEqual` or `APFloat::cmpUnordered` is returned for most of the non-normal cases, except when either of `f` or `g` is  not-[finite](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/APFloat.h#L380) or `Zero` for which the `APFloat::compare`  returns compare results other than `cmpEqual` and `cmpUnordered`. 

We handled all the non normal cases using 3 checks which are 
```
  if (cmpResult == APFloat::cmpEqual) return true;

  // For NaNs, one option was to return false, when compare result is `cmpUnordered`. We chose the path as
  // shown below to avoid false alarms. 
  if (cmpResult == APFloat::cmpUnordered) return f.isNaN() == g.isNaN();   


  if (!f.isFiniteNonZero() || !g.isFiniteNonZero()) return false;
```